cpp/src/arrow/compute/kernels/scalar_cast_nested.cc

```diff
@@ -404,14 +400,20 @@ struct CastStruct {
 
     int out_field_index = 0;
     for (int field_index : fields_to_select) {
-      const auto& target_type = out->type()->field(out_field_index++)->type();
+      const auto& target_field = out->type()->field(out_field_index++);
+      const auto& target_type = target_field->type();
       if (field_index == kFillNullSentinel) {
         ARROW_ASSIGN_OR_RAISE(auto nulls,
                               MakeArrayOfNull(target_type->GetSharedPtr(), batch.length));
         out_array->child_data.push_back(nulls->data());
       } else {
         const auto& values = (in_array.child_data[field_index].ToArrayData()->Slice(
             in_array.offset, in_array.length));
+        if (!target_field->nullable() && values->GetNullCount() > 0) {
```

Comment:
should we check `in_field->nullable()`?

Reply:
For correctness, I don't think this is needed (because the `values->GetNullCount() > 0` will take care of this), but for performance I could see it being nice to skip the `values->GetNullCount()` computation. I'll add the check.

Reply:
Yeah this might cause the null bitmap being counted, I think check nullable would be faster than checking this if not neccessary